plugins/filter/quote.py

```diff
@@ -0,0 +1,111 @@
+# Copyright (c) 2020 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import re
+
+from ansible.errors import AnsibleFilterError
+from ansible.module_utils.common.collections import is_sequence
+
+_UNSAFE_C = re.compile(r'[\s\t"]')
+_UNSAFE_CMD = re.compile(r'[\s\(\)\%\!^\"\<\>\&\|]')
+
+
+def _quote_c(s):
+    # https://docs.microsoft.com/en-us/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way
+    if not s:
+        return '""'
+
+    if not _UNSAFE_C.search(s):
+        return s
+
+    # Replace any double quotes in an argument with '\"'.
+    s = s.replace('"', '\\"')
+
+    # We need to double up on any '\' chars that preceded a double quote (now '\"').
+    s = re.sub(r'(\\+)\\"', r'\1\1\"', s)
+
+    # Double up '\' at the end of the argument so it doesn't escape out end quote.
+    s = re.sub(r'(\\+)$', r'\1\1', s)
+
+    # Finally wrap the entire argument in double quotes now we've escaped the double quotes within.
+    return '"{0}"'.format(s)
+
+
+def _quote_cmd(s):
+    # https://docs.microsoft.com/en-us/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way#a-better-method-of-quoting
+    if not s:
+        return '""'
+
+    if not _UNSAFE_CMD.search(s):
+        return s
+
+    # Escape the metachars as we are quoting the string to stop cmd from interpreting that metachar. For example
+    # 'file &whoami.exe' would result in 'whoami.exe' being executed and then that output being used as the argument
+    # instead of the literal string.
+    # https://stackoverflow.com/questions/3411771/multiple-character-replace-with-python
+    for c in '^()%!"<>&|':  # '^' must be the first char that we scan and replace
+        if c in s:
+            s = s.replace(c, "^" + c)
+
+    return '^"{0}^"'.format(s)
+
+
+def _quote_pwsh(s):
+    # https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_quoting_rules?view=powershell-5.1
+    if not s:
+        return "''"
+
+    # We should always quote values in PowerShell as it has conflicting rules where strings can and can't be quoted.
+    # This means we quote the entire arg with single quotes and escape the single quotes already in the arg.
+    return "'{0}'".format(s.replace("'", "''"))
```

Comment:
@jborean93 I know this technique works most of the time and has been used elsewhere in ansible (I forget if it's in the powershell shell plugin or directly win winrm/psrp), but it is possible to "single" quote PowerShell strings with characters other than the standard apostrophe `'` .

In PowerShell it's easiest to handle this with `[System.Management.Automation.Language.CodeGeneration]::EscapeSingleQuotedStringContent()` but since that's  not available, we can probably just hardcode the chars.

The really interesting thing about it is that the character delimiting the start and end don't even have to be _the same character_ as long ad they are both valid "single quote" chars.

For example consider this:
```powershell
’abc'
```
That's a valid single quoted string. Even though:
```powershell
"'abc'"[0] -as [int]
"'abc'"[-1] -as [int]
```
The code gen function handles it correctly:
```powershell
[System.Management.Automation.Language.CodeGeneration]::EscapeSingleQuotedStringContent("'abc'")
```

The [source for the code gen function is here](https://github.com/PowerShell/PowerShell/blob/b7cb335f03fe2992d0cbd61699de9d9aafa1d7c1/src/System.Management.Automation/engine/lang/codegen.cs#L21) but the really useful stuff for us is actually in the [CharExtensions class where they define the single quote characters](https://github.com/PowerShell/PowerShell/blob/b7cb335f03fe2992d0cbd61699de9d9aafa1d7c1/src/System.Management.Automation/engine/parser/CharTraits.cs#L18-L21) (these 4 are in addition to the standard apostrophe, [as seen in the `IsSingleQuote` function](https://github.com/PowerShell/PowerShell/blob/b7cb335f03fe2992d0cbd61699de9d9aafa1d7c1/src/System.Management.Automation/engine/parser/CharTraits.cs#L265-L272)).

The most likely place for this to bite someone is with non-English languages (this is actually how I found out about this, by mistake, forgetting to switch back to the English IME).

I think the key takeway from the code gen code is that those characters all need to be doubled against themselves, not changed to apostrophe. That prevents them from being used as string delimiters but ensures their content remains the same:
```powershell
[System.Management.Automation.Language.CodeGeneration]::EscapeSingleQuotedStringContent("won’t stop")
```

Reply:
Having a separate issue to track this would be great. I was unaware that chars other than `'` could be used as a quote and needed to doubled up in this scenario.

Reply:
Yeah it really caught me by surprise too; I think if I wasn't working on a powershell project that dealt with text in multiple languages I never would have run into it.

Opened #45 to track.